smbc: add extension for smbc-comics.com

solkaz opened the pull request:
Adds an extension for Saturday Morning Breakfast Comics

Closes #9648

Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

solkaz committed:
smbc: add extension for smbc-comics.com

Adds an extension for Saturday Morning Breakfast Comics

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,174 @@
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+class SaturdayMorningBreakfastComics : ParsedHttpSource() {
```
Extend `HttpSource` instead.

Also add a comment to say it is split from `HiveWorks`.

stevenyomi replied:
Should also add a comment in Hiveworks to say SMBC is split to separate extension.

solkaz replied:
As this is my first extension (and also my first time using Kotlin): can you explain why `HttpSource` is preferred over `ParsedHttpSource` in this case?

stevenyomi replied:
Shortly put, you're not using most of its methods, and we are deprecating that class whatever source website you're targeting. (I updated the contribution guide a few days ago)

If you want to know more, `ParsedHttpSource` is designed for traditional server-rendered websites where DOM structures and CSS selectors are very simple. Many websites now have some Ajax/API calls and complex page structures, so people usually need to override methods from `HttpSource`. Besides, there's really not much boilerplate code to write to achieve a `ParsedHttpSource` - basically `response.asJsoup().select(selector).map(toSManga)`, which is a lot more readable than a bunch of pre-defined abstract methods.

stevenyomi replied:
You should add comments before both classes to mention their link.

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,174 @@
+     *  the error here, it'll be treated as an any other failed request would; therefore we check
+     *  if the archive page specifically is being fetched, and change the error code if that's the case
+     */
+    private fun errorIntercept(chain: Interceptor.Chain): Response {
```
Can override `fetchChapterList` and use `asObservable` instead of `asObservableSuccess`. Then this is not needed.

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,174 @@
+        return pages
+    }
+
+    private fun parseDate(dateStr: String): Long {
```
Use `SimpleDateFormat.tryParse(String?)` provided by the `core` module.

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,174 @@
+    override fun fetchMangaDetails(manga: SManga): Observable<SManga> = Observable.just(manga)
+
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val chapterList = super.chapterListParse(response).distinct()
```
Is there any duplicate items? The behavior of calling `distinct()` on `SManga` is undefined.

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,174 @@
+        val chapterList = super.chapterListParse(response).distinct()
+        return chapterList.mapIndexed {
+                i, ch ->
+            ch.apply { chapter_number = chapterList.size.toFloat() - i }
```
Behavior looks weird. Should number the chapters from 1 then reverse the list. Fix it after you extend `HttpSource`.

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/build.gradle:
```diff
@@ -0,0 +1,8 @@
+ext {
+    extName = 'SaturdayMorningBreakfastComics'
+    extClass = '.SaturdayMorningBreakfastComics'
+    extVersionCode = 3
```
```suggestion
    extVersionCode = 1
```

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/build.gradle:
```diff
@@ -0,0 +1,8 @@
+ext {
+    extName = 'SaturdayMorningBreakfastComics'
```
```suggestion
    extName = 'Saturday Morning Breakfast Comics'
```

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,174 @@
+
+        const val IMAGE_SELECTOR = "img#cc-comic"
+
+        const val AFTERCOMIC_SELECTOR = "#aftercomic > img"
```
Inline these.

stevenyomi reviewed (commented):
Should remove all SMBC-related logic from Hiveworks and throw an exception (in their `chapterListRequest`) to let users download this new extension and migrate. And bump Hiveworks too.

Also remove the code mentioned https://github.com/keiyoushi/extensions-source/pull/9003#discussion_r2144166320.

cc @raxod502

solkaz committed:
hiveworks: remove references to Saturday Morning Breakfast Comics

Removes code that was made to handle reading SMBC specifically. If a
user still has the comic in the Hiveworks extension, they'll get a
warning to migrate to the SMBC extension.

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,146 @@
+            thumbnail_url = "https://fakeimg.ryd.tools/550x780/ffffff/6e7b91/?font=museo&text=SMBC"
+        }
+
+        return Observable.just(MangasPage(arrayListOf(manga), false))
```
```suggestion
        return Observable.just(MangasPage(listOf(manga), false))
```

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,146 @@
+                    response
+                }
+            }
+            .map { response ->
```
```suggestion
            .map { response ->
                if (!response.isSuccessful && response.code != 500) {
                    response.close()
                    throw Exception("HTTP ${response.code}")
                }
```

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,146 @@
+                    pages.add(Page(1, "", wordWrap(image.attr("title")).image()))
+                }
+                pages.add(Page(2, "", document.select("#aftercomic > img").attr("abs:src")))
+                pages
```
Use `pageListParse()`.

The default `pageListRequest()` is already what you're using (in fact, `baseUrl + chapter.url` which is more correct) so you can omit that.

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,146 @@
+            artist = "Zach Weinersmith"
+            author = "Zach Weinersmith"
+            status = SManga.ONGOING
+            url = "comic/archive/"
```
```suggestion
            url = "/comic/archive"
```

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,146 @@
+    override fun fetchMangaDetails(manga: SManga): Observable<SManga> = Observable.just(manga)
+
+    override fun fetchChapterList(manga: SManga): Observable<List<SChapter>> {
+        return client.newCall(GET("$baseUrl/${manga.url}"))
```
```suggestion
        return client.newCall(chapterListRequest(manga))
```

stevenyomi commented on src/en/saturdaymorningbreakfastcomics/src/eu/kanade/tachiyomi/extension/en/saturdaymorningbreakfastcomics/SaturdayMorningBreakfastComics.kt:
```diff
@@ -0,0 +1,146 @@
+                        val (date, title) = element.text().split(" - ")
+                        chapter.name = title
+                        chapter.date_upload = dateFormat.tryParse(date)
+                        chapter.chapter_number = index.toFloat()
```
```suggestion
                        chapter.chapter_number = (index + 1).toFloat()
```
